### ISSUE TITLE
MGD2 1756K

### ISSUE THREAD

**Issue by ronnybkk** [keirf/FlashFloppy#134]
Hello Keir,

I have yet another weird IBM format unsupported by FF.
It's a 1756KB image produced by another obscure game copier: the Bung Multi Game Doctor 2
Only the MGD2 can format 1756K floppies but other products manufactured by the same company can read those (Game Doctor SF3, SF7...)
When you format a disk in this device you have the option of formatting "ultra density" floppies
As quoted from the original chinglish documentation:

"(2) FU (1.756M Byte i.e. 13.5M Bit disk format):  Follow the
    flow chart to perform.  This is very useful for new disks
    or disks which have error.  This format allows a disk to
    hold thirteen 1M game."

I'd gladly replace my dead drive with FF if it can support it.
Luckily I managed to dump of a floppy formatted by the MGD2 here:
http://s000.tinyupload.com/?file_id=94384535772682066818

When I open this image on Winimage, I can see the files but I get the wrong filesize, the same way I do when I open up the content of the floppy on the PC.
So that make me think the dump is correct, although it was reported by HxCFloppyEmulator v2.1.11.1 having 1774 sectors

1756 is a weird number as I would expect 80Tx2Sx22SPT = 1760K
But unfortunately 1760K images aren't not read by the device and when I try the 1756k dump on 0.9.21a I get an error 31

**Comment by keirf** [comment]
I don't think 1756k makes any sense, but 1760k would be 22spt as you say.

**Comment by keirf** [comment]
I have added a 1760k format to the firmware, and here is a test build for you to try. I will state thye obvious: it needs a 1760k IMG file, not 1756k ;) So maybe pad with zeroes if you have the latter?

[ff_134_1.zip](https://github.com/keirf/FlashFloppy/files/2118183/ff_134_1.zip)

**Comment by ronnybkk** [comment]
Thanks again for your prompt and efficient help ;)

Unfortunately the upgrade didn't work
I know how weird that 1756K format sounds but this is exactly what is written on the GDSF7 menu (cf. pic) when I insert one of these disks. This is also exactly the size I get when I create a dump of the floppy with HxCFloppyEmulator.

The only reference I found on the net referring to that peculiar 1756K size is here:
https://manned.org/xdfcopy/7c98337f
But then again it refers to xdf which is 80Tx2Sx23SPT = 1840K. I tried and it didn't work either.

Now I noticed something that may give us a clue:
The GDSf7 has no problem reading the disk and launch the game, but when it tries to copy it, it fails at the 41st track.
When I dump the floppy on HxCFloppyEmulator it only sees it as a 1774 sectors floppy which seems to be half of the capacity.
But if the floppy is 1774x2 with 80 tracks, then it means there are 22.175SPT...is that even possible?

Sorry I can not come up with more info, I'll try to experiment a bit more.
Would dumping that floppy with a Kryoflux be the only way to know for sure what we're dealing with?

![1756k](https://user-images.githubusercontent.com/36294597/41646113-2b7eb6c6-749d-11e8-868f-226d70d20be2.jpg)

**Comment by keirf** [comment]
Yes a Kryoflux style dump may be needed. Like XDF it may involve packing each track with multiple sectors of different sizes.

**Comment by ronnybkk** [comment]
Ok I managed to get it working by making a new dump in hfe this time. Loading times are extremely slow though, so if you can figure out more info from it so we can have support in IMG, that'd be fantastic:
http://s000.tinyupload.com/index.php?file_id=06670287391459318709

**Comment by keirf** [comment]
The format is tracks 0 and 1: regular IBMPC HD (18 * 512b sectors). All other tracks are 11 * 1024b sectors with inter-sector gap = 0. No interleave. No skew.

**Comment by keirf** [comment]
Good news is I can probably support this not too much hassle. Otoh perhaps you are the only person in the world who cares, and is there a standard image format or anything like that?

**Comment by keirf** [comment]
Regarding HFE being slow, FlashFloppy needs further optimisation for the high-density HFE case. However it can work reliably today, it's just fussy about the USB stick. Perhaps you could try a different USB stick?

**Comment by ronnybkk** [comment]
Thanks for taking the time to look into it Keir.
The other track you mean 2x1024b so that makes 18x512bx2x80 (1,474,560b) + 2x1024x2x80 (327,680b) so a total of 1,802,240b, right?

I'd be lying if I was telling you that many people would be interested in it ;)
But that would make FF the ultimate replacement solution for game copiers FDD. The only one left to test is the Super Wildcard DX2 ED 2.88MB but I saw that FF already supports it.

For the record a similar machine was used to dump most of the GB, GG and PCE roms available on the net today. I own one of the rare dumping PCB card and I'm working with dbjh from Ucon64 to support many of these old copiers.
All I need is a way to fully support all the functionalities of the original FDD, so basically a 1756K blank image file I can read and write to. I wouldn't even know how to create one of these actually since the dump I made can not be modified via existing softwares. Unless formatting on the MGD2 a valid IMG/HFE works.

I'll try a different USB stick to see if things improve with HFE

Thanks again for your precious help, we made some progress ;)

**Comment by keirf** [comment]
No I mean 18*512b*2 + 11*1024b*2*79 = 1,798,144b = 1756kb exactly

**Comment by keirf** [comment]
Any luck with HFE?

**Comment by ronnybkk** [comment]
Hello Keir, I tried several USB keys (sandisk,lexar...) but always get slow loading times

**Comment by keirf** [comment]
How much too slow, is it consistent and is it consistent across USB sticks? Perhaps you could attach an HFE image for me to look at, to this ticket?

What is the minimum extra functionality you need to get yourself working okay (even if slow) with FlashFloppy? Is working with HFE ok? Or do you really need IMG?

**Comment by ronnybkk** [comment]
I'm not at home till Wednesday but I can time it then if it helps. I'd say maybe it takes 30secs to load 1MB in IMG and 3 mins in HFE. And yes timings are consistent across Usb sticks.

The HFE image I posted earlier that you analyzed to determine its format is the one I've been making all my tests with.

Another inconvenience of the HFE format is that it takes more than twice the size of an IMG
For instance that 1756KB image is 3,893,760 bytes in HFE. The same goes for regular 1440KB floppy images in HFE.

I think it should work as far as reading even if it's slower with a HFE image. As for writing I would need to make a 26p-34p adapter and install it on the MGD2, this machine uses a slim laptop FDD.
I'll try to make one of those asap
So far I've done the tests on the GDSF7 which can read but can't write on 1756KB floppies.

**Comment by sphiend** [comment]
Hey guys.  +1 for someone that would love to have a working floppy emulator replacement for my MGD2s.  Rest assured that you are not the only one in the world that cares. 

I have one original unmodified MGD2 Drive unit and a number of units that were modified (by someone else) to work with an NEC slim drive.  My one attempt a long time ago to get a standard gotek working was unsuccessful.  I also tried a 26-34 pin adapter made for a Yamaha keyboard without success.  

Let me know I there is anything I can do to help.  Most of the discussion above is Greek to me.  Sounds like you are making image files of the original floppies using a tool (HFE).  I’d need some hand holding but could maybe provide something useful.

I provided Edoardo with a unit to map the pins.  Hopefully the board he is designing will allow for use of original 26 pin cable rather then the 34 pin hack.

**Comment by keirf** [comment]
Please try the firmware below, which should correctly handle .IMG files of size 1756kB as MGD2 image files:

[flashfloppy-v1.0-mgd2.zip](https://github.com/keirf/FlashFloppy/files/2832426/flashfloppy-v1.0-mgd2.zip)

**Comment by ronnybkk** [comment]
Keir, you rock! I'll give it a try tomorrow. Would you be kind to also create a blank 1756KB IMG file as HxCFloppyEmulator can not create those, neither can WinImage.
This way I should be able to see if I can read/write on the GDSF7

@sphiend 
Thanks for reviving this thread ;)
I've been working with dbjh from uCON64 on various things regarding copier support lately (most of which can be find on Tototek forum) but I was meaning to resume testing on the MGD2 at some point. It's nice indeed to know I'm not the only one who cares about this machine ;)
First let me resume the progress so far: Flash Floppy is fantastic for copiers since most of these use standard IBM PC format IMG/IMA files generated by WinImage.
The MGD2 can read/write 1.44MB/1.6MB so that should work fine if you decide on using a Gotek with FF instead of  your current floppy drive.
The MGD2 can also read and create 1756KB IMG files which can also be read by the GDSF3/6/7.
The whole purpose of that thread is to support that format to have a FULL support of the MGD2.
Regarding the installation of the Gotek with the 34-26 conversion, I don't see why it wouldn't work but hopefully I'll have time to give it a try soon. I also have a GDSF2 that needs it since it's using the exact same drive as the MGD2...and both are damaged.

**Comment by keirf** [comment]
Here you go. It's just an empty file of zeroes of the correct length.

[mgd2.img.zip](https://github.com/keirf/FlashFloppy/files/2833089/mgd2.img.zip)

**Comment by sphiend** [comment]
ronnybkk,

I have a gotek for this project but I’m waiting on parts to mod it with flashfloppy.   

I was given a pdf with the MGD2 pinout so that night help if you run into issues.

One thing to note is that if your MGD2 is stock, it will have con3 closed to supply the 5 volts through the 26 pin cable.  If your drive was ever replaced with a 34 pin drive, they likely opened it up at con3 to cut power through the cable and then pulled 5v from somewhere else on the boat for the 34 pin drive.  

So what program do I need to get to make some images of my existing MGD2 floppies? (Is it WinImage?)  I’d have to guess these were not using the ultra density mode as most were 8 or 4 Mbit.

**Comment by ronnybkk** [comment]
Thanks a lot Keir! That was fast ;)
May I ask you how you created it? I'm considering making a software to ease the conversion process later but couldn't find much documentation on the IMG format

@sphiend 
Thanks but I think I have that PDF explaining how to install a regular 34 pins FDD.

To make 4/8/12 Mb floppies you can use the 1.44M and 1.6M floppy images attached.
Not sure if those are empty but basically you just open them with Winimage, inject your roms then save.
For the Ultra density I'll run some tests tomo and see how it goes but I doubt Winimage will enjoy that exotic format, hence the reason I'd like some more documentation on the IMG format so eventually we can create our own.
[floppies.zip](https://github.com/keirf/FlashFloppy/files/2833288/floppies.zip)

**Comment by kinmami** [comment]
[MGD2.pdf](https://github.com/keirf/FlashFloppy/files/2835116/MGD2.pdf)

in attachment,
the pinmap of MGD2 FDD connector onboard,
where pins 7~13 are kept floating (originally connected to +5V) to be compatible with flashfloppy/gotek drives,
pins 2-4 not connected,
pin 12 and 34 connected each other and to READY signal provided by 74LS244 IC mounted on MGD2

flashfloppy/gotek drive requires S0 jumper open and S1 jumper closed in order to avoid error on MGD2

**Comment by kinmami** [comment]
using flashfloppy-v1.0-mgd2.zip,
so far MGD2 does display the floppy drive icon
but flashfloppy/gotek drive is not loading tracks on each of proposed IMG files

**Comment by ronnybkk** [comment]
Hello Kinmami, thanks for your contribution.
Here I attached the PDF regarding the mod. I've never performed it but nothing fancy really, just add the missing pins and tap the +5V.
So you're saying even a regular 34 pins FDD doesn't work with that mod and need additional modifications or it just concerns the Gotek?
I realized the MGD2 supports 720K (FL), 1.44MB (FH) and 1.75 MB (FU) so only the 1.44MB image I provided above should work.
I'll try to find some time today to run some tests on the GDSF7 to cehck if it can at least read the 1756K IMG.
[MGD2_-_How_to_replace_floppy_drive_by_BAD.pdf](https://github.com/keirf/FlashFloppy/files/2835241/MGD2_-_How_to_replace_floppy_drive_by_BAD.pdf)

**Comment by kinmami** [comment]
Hello ronnybkk, thanks
in fact it is what I did actually but I cannot see the flashfloppy/gotek drive loading IMG @ 1.44M

**Comment by kinmami** [comment]
what about jumpers on your FDD?

**Comment by keirf** [comment]
> Thanks a lot Keir! That was fast ;)
> May I ask you how you created it? I'm considering making a software to ease the conversion process later but couldn't find much documentation on the IMG format

The format appears to be regular PC HD on tracks 0 and 1 (ie. cylinder 0): 18 sectors, 512 bytes per sector, numbered from 1 upwards
All other tracks have: 11 sectors, 1024 bytes per sector, no (or very tiny) inter-sector gap.

Those ultra-density tracks have to be written/formatted track at a time I believe, as the inter-sector gap is too small to safely overwrite sector-at-a-time.

**Comment by keirf** [comment]
I would get PC HD (1.44) images working with release firmware (v1.0) first, then update to the experimental firmware above (v1.0-mgd2) to test HD and UD. You can downgrade/upgrade/crossgrade between versions as you like using the same "firmware update" process if you don't want to reflash via DFU/serial.

**Comment by ronnybkk** [comment]
@kinmami Jumpers? you mean s0 or s1? It should be s1 but I'll check on my FDD
BTW in case it's useful for you, here are the error codes:
E1:  error in disk format
E2:  cannot find the required file
E3:  incorrect file length
E4:  error in disk (bad disk?)
E8:  error in the index file, 'MULTI-GD'
E10: cannot find the index file, 'MULTI-GD'
E14: disk index full, cannot write new file
E15: insufficient disk space for writing new file
EIF: error in verification

@keirf 
Thanks, good to know.
Regarding the creation of those IMG, I meant what tools are you using? hxfce, dd or another linux tool?

**Comment by keirf** [comment]
To create the totally empty zeroed 1756kB IMG file:
```
dd if=/dev/zero of=mgd2_1756.img bs=1024 count=1756
```

**Comment by keirf** [comment]
More explanation: IMG is just a dump of logical sector contents -- FlashFloppy infers sector sizes, sectors per track, etc (the geometry of the disk) from the IMG file size. It uses FF.CFG values (specifically host= setting) to disambiguate further, where file size is insufficient.

So a blank IMG can simply be a zeroed file of the correct size: it may then need logical formatting on the host system (eg. to put a FAT filesystem on it).

**Comment by sphiend** [comment]
Had similar experience as kinmami with test firmware on my Amiga gotek.  Pulled images off some actual MGD2 disks using winimage.  (1.44 meg) Renamed the .IMA to .IMG since the Internet assured me they are same thing.

Put gotek jumper on JC.  Tried it both with S0 open and closed.  Closed, I got E4.  Open, I got disk icon with no track reading.  I didn’t think to try it with S1.

**Comment by keirf** [comment]
I think ronnybkk has had success with 1.44m IMG files? Let's see what he says.

**Comment by sphiend** [comment]
To be clear, I think we are working on the same file system on two different devices.  Though, I may have misunderstood.  Apologies if that is the case.

My understanding is that to date, ronnybkk has been using a gotek with FF on a Game Doctor SF7.  This hardware is capable of reading the older MGD2 formats, including the ultra high density formatted disks.

I believe he has an original MGD2 drive with a dead FDD but it has not yet been modified (hacked) to use a 34 pin drive and his original 26 pin Mitsubishi is a goner. (As they all are)  The PDF document he posted is instructions for how to solder in headers and pull power to use a 34 pin replacement.  For the testing to date, I believe he has not been using the MGD2 original unit.  Please correct me if that’s wrong.

Both kinmami and myself are working off an original MGD2 drive unit that has been modified to accept a 34 pin NEC drive.  My replacements have all been NEC 1138H models but I’ve seen reference to NEC 1148H being used (that’s what’s talked about in the PDF). I’ve never gotten an 1148H to work but the ones I’ve bought might just be bad.

**Comment by ronnybkk** [comment]
I had success with 1.44MB and 1.6MB on all the copiers. I haven't tested the MGD2 coz I still need to make the 34pins adapter as sphiend mentioned.
The closest thing to it that I have working is the GDSF7 from the same family, hence why I'm running tests on it. Also it provides more visual feedback than the poor MGD2 2 characters led display.

So I tried fw 1.0 and it was throwing ERR31 at everything, on the other hand, fw 1.0 mgd2 was at least trying to access the disk. I got quite a few 1756K IMG from previous tests but I can't recall what they are...some of them can read the TOC and see it's a 1756K disk but stop loading right at the beginning.
The empty 1756K image is rejected and keep trying to access the first boot sector.

The only valid 1756K image that works is the attached HFE dump I did a while ago, but again loading is very slow.
If we could someway convert that image to a valid 1756K IMG, I could try to hexedit the file to understand the structure.

@sphiend 
Can you send me your MGD2 dump of that 1.44MB disk, see if it works on my GDSF7?
Have you tried to use a regular 34p FDD instead of the Gotek?
Actually I have the rare backup interface for the MGD2, so I could try different things once I can find time to modify the MGD2.
Problem is timing is not really good at the moment as I'm currently building a bar in Bangkok...so you're all welcome for a drink when it's done ;)

[MGD2 1774 sectors 1798144 bytes.zip](https://github.com/keirf/FlashFloppy/files/2836637/MGD2.1774.sectors.1798144.bytes.zip)

**Comment by keirf** [comment]
You would expect FW v1.0 to throw ERR31 on any 1756k IMG as that image size is unknown to it. If you were to try 1.44M IMGs with v1.0 then that should of course work fine.

I will take another look at the HFE dump.

**Comment by keirf** [comment]
The HFE format is as I recalled. So the 1756k IMG handling I added should work, assuming it's not got a bug ;)

I attach a conversion of your working HFE to IMG format. Could you try it?

[MGD2 1774 sectors 1798144 bytes_hfe.zip](https://github.com/keirf/FlashFloppy/files/2837578/MGD2.1774.sectors.1798144.bytes_hfe.zip)

**Comment by ronnybkk** [comment]
Thanks for all you efforts Keir ;)
Unfortunately it doesn't work, I see the file, it's a 1756K disk but it freezes on the first track

**Comment by sphiend** [comment]
@ronnybkk,

Shoot me an email address you want some images sent to.

**Comment by ronnybkk** [comment]
@sphiend 

You can attach one here or send it to ronnybkk@hotmail.com

**Comment by keirf** [comment]
> Thanks for all you efforts Keir ;)
> Unfortunately it doesn't work, I see the file, it's a 1756K disk but it freezes on the first track

That's a shame :( Did 1.44M images work okay with that firmware (1.0-mgd2)?

**Comment by ronnybkk** [comment]
Yes, 1.44M IMG work.
One thing I wondered though, you said you created the empty IMG with
dd if=/dev/zero of=mgd2_1756.img bs=1024 count=1756

I don't see where T0 and T1 are 512B
You were saying earlier the disk image should be
18x512bx2 + 11x1024bx2x79 = 1,798,144b = 1756kb

**Comment by keirf** [comment]
That's just a convenient command line to create a 1756kB file: the bs=1024 has nothing to do with the sector sizes implicit in that image file when FlashFloppy parses it.

**Comment by sphiend** [comment]
@ronnybkk

Done.

**Comment by keirf** [comment]
I'll have to investigate why the 1756k IMG didn't work. It's a bit surprising, I will have to do my own test.

**Comment by ronnybkk** [comment]
> 
> 
> That's just a convenient command line to create a 1756kB file: the bs=1024 has nothing to do with the sector sizes implicit in that image file when FlashFloppy parses it.

Got it ;)

@sphiend 
Didn't get anything, not in my spam either

**Comment by sphiend** [comment]
@ronnybkk

Not sure what went on.  Tried again.  See if that worked.

**Comment by ronnybkk** [comment]
@sphiend 
Nope, still nothing. Attach the file here, just drag and drop it

**Comment by keirf** [comment]
Just an update, I will soon extend IMG.CFG to allow configuration of mixed sector sizes, both across and within tracks. When this is done, it should be easy to make a configuration for MGD2.

**Comment by ronnybkk** [comment]
Great Keir! You rock ;)
I'll make sure to find some time to test it properly

**Comment by sphiend** [comment]
So I’ve been wondering if the problem with reading of disks using actual MGD2 hardware has to do with the READ data signal being too slow to rise (from a FDD emulator).  I know Edoardo was unable to get it working on his end.

This is a 26 pin drive similar to those in the FM Towns Marty.  In that’s case they installed a pull-up resistor to correct the slow timing.

described here.

https://www.jammarcade.net/using-the-hxc-floppy-emulator-with-a-fujitsu-fm-towns-marty/

**Comment by keirf** [comment]
Did no-one have success even with HFE images?

**Comment by sphiend** [comment]
I  can’t speak for ronnybkk; however, I believe he was using other copiers that support the MGD2 format.  I do not believe he was using an actual Bung Multi Game Doctor.

I sent Edoardo an MGD2 unit and he was unable to get it to read any disks using a floppy emulator.  He was kind enough to map out the pins.  

My suspicion is that there is something similar to the Marty issue that will require an adapter (with some additional resistance) to get the signals recognized by a FDD emulator.  (When it works fine with a physical drive)

Here is an example adapter that someone makes and sells for the Marty.  

https://rover.ebay.com/rover/0/0/0?mpre=https%3A%2F%2Fwww.ebay.com%2Fulk%2Fitm%2F312816082135

**Comment by Stuckinthe1990s** [comment]
Sort of off topic. Does anyone know the specs on the drive belt for the Mitsubishi floppy in the mg2 and possibly a place to buy one?

**Comment by keirf** [comment]
Okay, finally IMG.CFG should support MGD2 image files! I attach a test firmware for you to try. You need to copy `examples/Host/Bung_MGD2/IMG.CFG` to the root of your USB stick (or `FF/` if you have that subfolder on your USB stick). In the same folder there is also an example empty image file of the correct size (exactly 1756kB).

Particularly anyone who has had success with HFE image files: I'm very interested if IMG/IMA now works too!

[ff_134_2.zip](https://github.com/keirf/FlashFloppy/files/5358929/ff_134_2.zip)

**Comment by sphiend** [comment]
Thanks Kier.  I’ll see if I can update my drive this weekend and hook it up to an actual MGD2 unit.  Been quite a while since I tried this.  My drive is currently configured (jumper wise) for IBM PC use.  Will I need to reconfigure the hardware jumpers?  If not, I assume just flash the image and drop the config file on there.

-Doug

**Comment by keirf** [comment]
Here's a further revised version which should make the track format even closer to proper MGD2 (now GAP3 is 0 as it should be). All instructions as for ff_134_2.

[ff_134_3.zip](https://github.com/keirf/FlashFloppy/files/5359168/ff_134_3.zip)

**Comment by keirf** [comment]
@sphiend The jumpers depend on exactly what drive type the MGD2 (via any adapter you are using) expects. Assuming interface is not specified in an FF.CFG file, you will want JC jumpered for IBM interface (pin34=DSKCHG) or JC open for Shugart (pin2=DSKCHG,pin34=READY). And you will want one of S0 or S1 depending on whether drive select is pin 10 or pin 12.

Looking at @kinmami post long ago where he attached MGD2.pdf, it looks like you may want to jumper S0 only. This sets the Gotek up as a bog standard Shugart drive responding on drive-select pin 10.

**Comment by ronnybkk** [comment]
Wow, amazing work once again @keirf . I'll do my best to work on it next week. Thanks for your efforts ;)

**Comment by sphiend** [comment]
Great news Kier.  You have done it!

It is working on one of my MGD2s that has been modified to work with a 34 pin drive.  It was common for the original 26 pin slimline Mitsubishi to be replaced with a NEC 34 pin drive by adding some additional headers, crimping a new cable, and pulling some power from the PCB.  I have not yet tried it with an original 26 pin unit.  (Will if I get a chance)

To your point, configuration is:
JC - Open
S0 - Closed
S1 - Open

I tried it using both a standard gotek and Edo’s dedicated FlashFloppy board.  Both appear to work fine.  I’m hopeful the FF board, with its built in 26 pin header, will be a direct fit for a stock MGD2.

I used actual IMG files that I made when this discussion first started.  I believe they are 1.44 meg images.  Thus, I can’t say if your larger file works.  I’m not sure I even could test it if it doesn’t have a loadable file and the index file within your image.  (It may, I didn’t look)

I made a few videos for you.  The unit is currently not in the case..

Loading image — standard gotek 

https://youtu.be/NxMspWib-IQ

Loading image - FlashFloppy drive

https://youtu.be/xAk928yVNQ0

Testing flashed image on real hardware 

https://youtu.be/6oYk-PPTwmk

Thanks again for all you have done!  

Regards

**Comment by Stuckinthe1990s** [comment]
If you need help formatting and loading files or anything I can help with I got mine up and running too. 
https://youtu.be/uzgRoCENyV0
Mgd2 
--
Sent from myMail for Android Saturday, 10 October 2020, 04:34p.m. -04:00 from sphiend  notifications@github.com :

>Great news Kier.  You have done it!
>
>It is working on one of my MGD2s that has been modified to work with a 34 pin drive.  It was common for the original 26 pin slimline Mitsubishi to be replaced with a NEC 34 pin drive by adding some additional headers, crimping a new cable, and pulling some power from the PCB.  I have not yet tried it with an original 26 pin unit.  (Will if I get a chance)
>
>To your point, configuration is:
>JC - Open
>S0 - Closed
>S1 - Open
>
>I tried it using both a standard gotek and Edo’s dedicated FlashFloppy board.  Both appear to work fine.  I’m hopeful the FF board, with its built in 26 pin header, will be a direct fit for a stock MGD2.
>
>I used actual IMG files that I made when this discussion first started.  I believe they are 1.44 meg images.  Thus, I can’t say if your larger file works.  I’m not sure I even could test it if it doesn’t have a loadable file and the index file within your image.  (It may, I didn’t look)
>
>I made a few videos for you.  The unit is currently not in the case..
>
>Loading image — standard gotek 
>
>https://youtu.be/NxMspWib-IQ
>
>Loading image - FlashFloppy drive
>
>https://youtu.be/xAk928yVNQ0
>
>Testing flashed image on real hardware 
>
>https://youtu.be/6oYk-PPTwmk
>
>Thanks again for all you have done! 
>
>Regards
>

**Comment by sphiend** [comment]
Thanks!

So, for completeness, I also tried a stock MGD (unmodified/26 pin) with the FlashFloppy drive’s native 26 pin connector. Wasn’t able to get the MGD2 to power on when the drive was connected. In this case, the 7805 voltage regulator on the power board got incredibly hot so I disconnected and ceased trying before I blew something.  I validated it doesn’t get that hot under normal operation. 

I didn’t see any jumpers or config options specifically related to the 26 pin connector on the FF board. 

So, I think unless someone has an easy fix, use of FF with an original MGD2 unit will likely require the unit be modified to work with a 34 pin drive.

**Comment by keirf** [comment]
Those 26-way pinouts can vary so you may have shorted out the power line to ground.

I am particularly interested in 1756kB images, if anyone has any.

Alternatively can the blank IMG that I included be formatted and then verified by the MGD2?

**Comment by sphiend** [comment]
The unit and its functions are pretty cryptic.  I would assume it must have a format option but I don’t remember off hand how to use it.



There is a second external circuit board with cartridge headers that attached for “backing up” from a cartridge.  (Picture above)

In that case, you had to load a specific disk (also pictured)  and execute specific program numbers to perform actions.  For this use case, I’d have to image that disk for use with flashfloppy.  Regardless, I would need to do some research to see if I could find documents on the subject / figure it out.  Unfortunately, any docs or info on these is hard to come by these days and my memory from 28 years ago is pretty hazy.


IIRC, someone on this list was using another (later)  backup unit with a UI that supported reading / writing 1756kB MGD2 images.  I think their specific use case was the larger image.  If they can’t validate, I’ll look into using the backup board and software to try to validate it that way.  May just take me a bit to find the info and experiment.

Regards.

**Comment by sphiend** [comment]
So I braved the malware filled recesses of geocities and others to find some instructions for performing a backup.  As best I can tell, there is no “format” option in the cryptic backup software. I suppose it would format the disk during backup using whatever size was needed to hold the image.  Personally, I don’t remember ever seeing a image that was bigger than 1.44 MB.  

Best I could do would be to create a copy of the backup software as a disk image and move that to the USB stick.  Boot to that with the backup board attached then switch over to your 1756kB image as the target disk for the backup.

Assuming success, check FDD image size to make sure it’s still correct then try to load game off that image to SRAM cartridge. 

Let me know if you want me to give that a shot.

**Comment by keirf** [comment]
I think we can leave it to @ronnybkk who originally opened this ticket.

**Comment by ronnybkk** [comment]
There are 3 types of formatting from the backup utility for the MGD2: FL, FH and FU. FU formats floppies as 1756kb.The GDSF7 is backward compatible with that format but can only read those disks.Easiest way for me to test if the new FW is working is to run a 1756kb IMG on the GDSF7.Then later I'll try to read/write/format directly from a MGD2.On Oct 11, 2020 22:53, sphiend <notifications@github.com> wrote:

So I braved the malware filled recesses of geocities and others to find some instructions for performing a backup.  As best I can tell, there is no “format” option in the cryptic backup software. I suppose it would format the disk during backup using whatever size was needed to hold the image.  Personally, I don’t remember ever seeing a image that was bigger than 1.44 MB.  



Best I could do would be to create a copy of the backup software as a disk image and move that to the USB stick.  Boot to that with the backup board attached then switch over to your 1756kB image as the target disk for the backup.



Assuming success, check FDD image size to make sure it’s still correct then try to load game off that image to SRAM cartridge. 



Let me know if you want me to give that a shot.

**Comment by sphiend** [comment]
@Ronniebkk

Do you have a 1756kb IMG with MGD2 formatted software and index inside it?  I assume if not — maybe could be made manually with winimage or similar.  

I assume the file provided is just an empty unformatted IMG file.

If you need me to try to do what I described, let me know.

**Comment by Stuckinthe1990s** [comment]
I've tracked down a bunch of utils for the mgd2.  If there's anything you're looking for I'll try to help.
https://youtu.be/uzgRoCENyV0
This is a vid of my unit with a 26pin drive.

**Comment by sphiend** [comment]
Thanks.  Ronniebkk mentions formatting from the backup utility.  The instructions I found didn’t mention a format option.  Do you know what utility he is talking about — and if the standard backup utility, do you happen to know the procedure to format the disk?  I’m going to try to get a bootable IMG version of the backup utility made today.

**Comment by Stuckinthe1990s** [comment]
I'm away from home till later tonight I can send a zip file of everything I have if you have a way for me to get it to you.

I formatted the disks just with the windows util in XP. I used ucon64 to convert rom formats to mgd2 and to split later files between disks. I can also share a couple known working disk images too later.

**Comment by ronnybkk** [comment]
I've attached a dump of a 1756kb image on the first post but it's not valid anymore. I can repost it when I'll be at my place.I own the original backup card, and the original backup floppies from Bung. Look up on Tototek I believe I uploaded them there years ago (Mystic_Merlin), otherwise I'll repost them too.On Oct 11, 2020 23:28, sphiend <notifications@github.com> wrote:

Thanks.  Ronniebkk mentions formatting from the backup utility.  The instructions I found didn’t mention a format option.  Do you know what utility he is talking about — and if the standard backup utility, do you happen to know the procedure to format the disk?  I’m going to try to get a bootable IMG version of the backup utility made today.

**Comment by sphiend** [comment]
I used my backup card and backup floppy on an original MGD2 unit today to test this.  

The MGD2 refused to write to Keir’s example image.  (1756). It returned an E15 error.  According to a tototek post, that’s saying not enough space.  It should be noted that I can’t open his image with winimage either.  But I’m not claiming to be knowledgable on how these images work.

All that said, it had no problem writing to (backing up to) a blank 1440 IMA file that I created with winimage.  I was able to confirm the backup was successful.  So, that said — it looks like the new firmware is good for reading and writing up to standard 1.44 meg disks.  The 1756 kB file provided didn’t work out with my tools.  YMMV.  

Here are the error codes —


MGD2 error code: 
E1: error in disk format 
E2: cannot find the required file 
E3: incorrect file length 
E4: error in disk (bad disk?) 
E8: error in the index file, 'MULTI-GD' 
E10: cannot find the index file, 'MULTI-GD' 
E14: disk index full, cannot write new file 
E15: insufficient disk space for writing new file 
EIF: error in verification ”

**Comment by keirf** [comment]
@sphiend The problem might be that the 1756kB image file isn't 'formatted'. It's just entirely blank (zeroes). Since nothing but MGD2 and friends uses this weird disk geometry, they must be able to format to this capacity themselves. But quite possibly your write attempt expected to see the image already formatted.

Just a guess. :) Thanks for trying anyway!

I think for starters we'd need to see 1756kB writes working to a real HD floppy. And then swap in the Gotek. At least then the process to format and write would be known and working on the MGD2 side.

**Comment by Stuckinthe1990s** [comment]
I have no idea if any of this helpful.  I love what you're doing.  Here is a few examples that I've tested and used on disk and the program i used to create the correct format for a disk based mg2

[http://www.mediafire.com/file/78qeethmzxftdwg/mg2.zip/file](url)

**Comment by keirf** [comment]
I'm going to leave this open a while longer, and also include my example MGD2 IMG.CFG in the next release, if only because it is another nice example of an IMG.CFG file. If this ticket peters out I will close it in the next while, pending any further useful testing anyone can do on 1756k images.

**Comment by ronnybkk** [comment]
So I tried a couple of things with the latest firmware, mostly on the GDSF7 as I realized something I'll explained later for the MGD2.
The GDSF7 also gives more feedback so it's practical for testing. Bung implemented this format support in their other copiers for owners of the MGD2 so they could still play their collection on future revisions. Hence 1756K is supported by the GDSF3, the GDSF6 and GDSF7 but I believe it's only able to read it.

First of all, it works! The GDSF7 is a able to id the disk as a 1756K floppy in IMG format.
As I understand this IMG.CFG file is parsing files based on the filesize and you can manually change the attributes, neat!
The only issue is that it loads as fast as the hfe file. For instance on the real floppy it takes 24 seconds to read a 8Mbits file while it takes 3:30mn with FF.
My understanding is that FF is not at fault here, the dumping is. I used the HXC app to dump it in hfe then converted it in IMG so whatever went wrong during the dump was carried over in the conversion.
Another thing to note is the example.img file is not working, it's throwing a disk format error.

I did a raw dump and posted it in the link below as well as a couple of MGD2 utilities and the 8Mbits dumps in hfe and IMG.
WinImage is able to open the file but it reports incorrect name and filesize so I guess that to manually inject a file into a proper 1756 IMG, one would have to create a special tool or some advanced dd command?
https://drive.google.com/drive/folders/1HySWPOsF4cdkrssHFnO_VGVVxfcpFeWF?usp=sharing

So, coming back to the MGD2, as I mentioned earlier I own 2 units and I realized they might not be on the same firmware.
Indeed I was not able to read the 1756K floppy on the black unit while it was created on the white one years ago. I also realized the newer toolset with the "X02" extension was not working on the black unit, only the "F02" would.
My guess is that my white unit was updated with the 2.30 bios but I'm not willing to run the update on the black one yet as I'm willing to test and document those machines.

I don't know if the raw dump would help in any way provided it was done properly by the software, as to explain why loading is so slow, but at least now we can run those in a non bloated hfe format.

Next step would be to hook up FF to the MGD2 and run more tests but it's taking more time.

The MGD2 geocities page is still online btw:
http://www.geocities.ws/multigamedoctor/background.html

**Comment by keirf** [comment]
@ronnybkk what is really useful is a dump of a real floppy. Is that what is in the RAW.zip in your link? Dumped with Kryoflux?

Dump of a Gotek is less interesting. I know what that looks like. Or I can generate it myself ;)

**Comment by keirf** [comment]
What I want to do is inspect the real floppy dump very carefully and see if there is any way to tweak IMG.CFG to make FlashFloppy's geometry as close as possible. To avoid this super slow read time. It seems maybe the sectors should be interleaved, or tracks skewed.

**Comment by ronnybkk** [comment]
Yes Keir, it's a dump of the real floppy, dumped with the HXC tool

**Comment by keirf** [comment]
A real disk that runs at full read speed (24s)?

Okay I note that the 1024-byte sector tracks have a slightly slower data rate. Maybe that matters. Try unzipping the following IMG.CFG file to replace the one currently on your USB stick.

[IMG.CFG.zip](https://github.com/keirf/FlashFloppy/files/5377099/IMG.CFG.zip)

Other things to try if you're happy to play with the IMG.CFG file would be to remove the `rate=` line I've just added and instead increase `gap3=` to something like 30.

Geometry is entirely determined by IMG.CFG: If the IMG file works at all then it's a good conversion. It is up to FlashFloppy to 'frame' it up into a raw track so that MGD2 devices are happy to run at full speed.

**Comment by ronnybkk** [comment]
I used the HXCFloppyEmulator software, used the "floppy disk dump option" but I realized I selected 80 tracks, shouldn't it be 81?

I will try to play with the IMG.CFG gap option now and report. Any idea why your example.img didn't work based on the analysis of the raw dump?

**Comment by keirf** [comment]
@ronnybkk My example.img is blank (full of zeroes) it's not formatted so although the sector layout will be right, the sector contents are... nothing. So the device will not like that.

Another thing to try in IMG.CFG is to add a line at the end `interleave = 2`. You may find this increases the speed to only approx half that of a real floppy (eg 50 seconds-ish?).

**Comment by ronnybkk** [comment]
Yes! Interleave is doubling the speed! Almost there ;)

**Comment by keirf** [comment]
How fast is it now?

**Comment by ronnybkk** [comment]
44 seconds! I'm trying gap options at this moment ;)

**Comment by keirf** [comment]
44 seconds isn't bad! Let me say that interleave is a band-aid solution. You will never get down to the 24 seconds of the real floppy disk, which has no interleave.

This is because the MGD2 is almost certainly reading sectors sequentially: if they are sequential on the track and MGD2 can read each one in time, that's great: You read the whole track in one revolution (200ms). But it's terrible if it usually misses the start of the next sector: Then you need a revolution per sector! That's going to be approx 10x slower. Which is what you saw.

By applying interleave 2, sequential sectors have another sector placed between them. This guarantees the MGD2 will have enough time to prepare for that next sector. However, it obviously then means you need at least two revolutions to read all sectors sequentially (that's 400ms).

Does that make sense?

So `interleave = 2` isn't bad. But your tweaking should be done with `interleave = 1` because that is what will allow the read speed to suddenly jump to 24 seconds when we get it juuuust right.

**Comment by keirf** [comment]
Here are other things to consider, at the end of IMG.CFG:
```
gap3 = 30
rate = 500
gap4a = 200
interleave = 1
```

500kbps is the default HD data rate.

GAP3 is the gap between sectors. Larger gives more time for preparation to read that sector.

GAP4A is the gap at the start of the track after the index mark. Actually MGD2 tracks seem to be a bit shifted: Lots of gap at the start of the track, and then the last sector actually crosses the index mark and into the start of the track. Quite unusual! Probably because the write speed is slow (it's not 500kbps like we expect but more like 485kbps). Anyway, this larger gap4a tries to simulate the gap at the start of the track. Bit of a long shot.

Overall GAP3 is the biggest chance :+1:

**Comment by ronnybkk** [comment]
Keir, you're a freaking genius! We reached the 24s mark with 

gap3 = 30
rate = 500
gap4a = 200
interleave = 1

All your explanations do make sense, thanks for so much dedication and support! Fascinating really ;)

**Comment by keirf** [comment]
Okay, for my satisfaction, can you try removing all those lines except gap3?

**Comment by ronnybkk** [comment]
Yes it also reaches 24s with the following:
cyls = 80
heads = 2
tracks = 0
  secs = 18
  bps = 512
tracks = 1-79
  secs = 11
  bps = 1024
  gap3 = 30

When I tried before with gap3 = 30 and interleave = 2 it was not giving such results.
I guess with gap3 = 30 only, there's less trickery involved and more chance to make it work with other dumps?

Anyway, thanks a million again for this achievement. You have a nice bottle of wine waiting at the bar if you ever swing by Bangkok ;)

**Comment by keirf** [comment]
I explain above why interleave 2 could never reach the speed of properly configured interleave 1. :) Thanks!

**Comment by sphiend** [comment]
Keir, thanks again for all your efforts.  It is really nice to have this as an option.  I appreciate everything you do for the community.

Will all of the discovered IMG tweaks discussed be integrated and included into a future release or should I manually modify mine based on this thread?

To others in this thread — 

someone asked about the Mitsubishi belt.  I’ve no idea the answer but I would like to try to repair mine.  Perhaps we can start a thread on tototek or elsewhere.

**Comment by keirf** [comment]
@sphiend The updated IMG.CFG will be in the next FlashFloppy release (v3.20) under `examples/Host/Bung_MGD2/`

**Comment by keirf** [comment]
I should get the release done possibly tomorrow. We'll see!

**Comment by sphiend** [comment]
Thanks again!

**Comment by FireDVB** [comment]
Hello.

Does anyone have a floppy image with 1756K?
All of these are not working.
with a 1756K.img you cannot delete Tetris and Mgd2.
With another 1756K image from here you can copy to it, but there should still be something free, but it shows me 0 bytes free and a disk error.

Who can help me with this.
The images from ronnybkk "floppies.zip" work perfectly, only one image with 1756K is missing

Thanks

Greetings FireDVB

**Comment by FireDVB** [comment]
In the 90s I formatted normal MFD-2HD diskettes 1.44MB with VgaCopy 6.25 in DOS/Windows 95 with 1.75MB. It also works perfectly on the PC and also with the copy units.
What is the best way to get a 1.75MB floppy created and then as an img?
Nowadays very difficult to get these programs to run. Anyone tips?

Unfortunately, I don't have such old hardware to create a disc with 1.75MB with Vgacopy and then make an image out of it.
Maybe you have more options to test.

**Comment by philjay78** [comment]
> I have successfully created 1722K and 1785K floppy image. Works flawlessly.
> 
> Does anyone have a tip if I want to test with 2880K image how would the Img.cfg look like?
> 
> Thanks.
> 
> Greeting FireDVB

Hi, what image are you using to create the 1785 floppy?

**Comment by FireDVB** [comment]
I've attached the floppy images. Maybe someone needs it. I can't get the 2880K image to run with flashfloppy "Read Error".

[floppies.zip](https://github.com/keirf/flashfloppy/files/10832888/floppies.zip)
